Jerrod/Assignments/Rails_models/dojo_ninjas.txt

```diff
@@ -0,0 +1,118 @@
+| id | name | city | state | created_at | updated_at |
++----+------+------+-------+------------+------------+
+|    | test |      |       |            |            |
++----+------+------+-------+------------+------------+
+Dojo2.save
+   (0.1ms)  begin transaction
+   (0.1ms)  rollback transaction
+Dojo2.errors.full_messages
+ => ["State is the wrong length (should be 2 characters)"] 
+
+	make sure validations work
+
+FOURTH***************************************
+Dojo.first.ninjas.create(first_name: 'first', last_name:'last')
+Dojo.first.ninjas.create(first_name: 'Louis', last_name:'Tom')
+Dojo.first.ninjas.create(first_name: 'John', last_name:'Johnson')
+	create three ninjas that belong to the first Dojo
+
+FIFTH***************************************
+Dojo.second.ninjas.create(first_name:'Sam', last_name:'Winchester')
+Dojo.second.ninjas.create(first_name:'Dean', last_name:'Winchester')
+Dojo.second.ninjas.create(first_name:'Adom', last_name:'Winchester')
+	create three new ninjas that belong to the second dojo
+SIXTH***************************************
+Dojo.third.ninjas.create(first_name:'Michael', last_name:'ArchAngel')
+Dojo.third.ninjas.create(first_name:'Lucifer', last_name:'ArchAngel')
+Dojo.third.ninjas.create(first_name:'Gabriel', last_name:'ArchAngel')
+	create three ninjas for the third dojo
+SEVENTH***************************************
+Dojo.first.ninjas.all
++----+------------+-----------+---------+-------------------------+-------------------------+
+| id | first_name | last_name | dojo_id | created_at              | updated_at              |
++----+------------+-----------+---------+-------------------------+-------------------------+
+| 1  | first      | last      | 4       | 2016-12-02 03:08:59 UTC | 2016-12-02 03:08:59 UTC |
+| 2  | Louis      | Tom       | 4       | 2016-12-02 03:09:16 UTC | 2016-12-02 03:09:16 UTC |
+| 3  | John       | Johnson   | 4       | 2016-12-02 03:09:26 UTC | 2016-12-02 03:09:26 UTC |
++----+------------+-----------+---------+-------------------------+-------------------------+
+Dojo.second.ninjas.all
+Dojo.third.ninjas.all
+	Know how to get all of the ninjas for any dojo
+
+EIGHT***************************************
+Dojo.second.ninjas.select(:first_name).order("created_at DESC")
+	retrieve the first_names of the ninjas that belong to the second dojo and order the result by created_at DESC order
+
+NINTH***************************************
+	Delete the second dojo. How could you adjust your model so that it 			automatically removes all of the ninjas associated with that dojo?
+	**************FIRST WAY I DID THIS**********
+Dojo.second.ninjas.destroy_all
+Dojo.second.destroy
+	Delete the second dojo. How could you adjust your model so that it 			automatically removes all of the ninjas associated with that dojo?
+
+	One way is shown above, first destroy all the ninjas associated with it. 		Delete_all only deletes the reference, so you have to do destroy_all to 		delete the actual ninja
```

Comment:
delete_all would have deleted the actual ninjas as well.  one of the differences is that delete_all will delete all of the associated ninjas in one query, whereas destroy_all will first load up all the associated ninjas and then delete them one by one.  Here's the queries that are run with a delete_all:
<img width="762" alt="screen shot 2016-12-03 at 3 00 34 pm" src="https://cloud.githubusercontent.com/assets/18199508/20862167/52aa61b4-b969-11e6-9822-4ab68d79fde5.png">
And here's the queries run using destroy_all (notice that the console actually displays the objects that are destroyed as well whereas they weren't with 'delete_all'):
<img width="729" alt="screen shot 2016-12-03 at 3 04 42 pm" src="https://cloud.githubusercontent.com/assets/18199508/20862182/ec4e5582-b969-11e6-86a4-192fd02f02fb.png">

There's also one other difference that I'll note in the next comment